reader.js

```diff
@@ -7,15 +9,156 @@ const CLOSE_TIMEOUT = 10000;
+
+    if(request.method === 'GET') {
+
+        let requestQuery = new URL(request.url, 'https://habmc.stanfordssi.org/');
+
+        if(requestQuery.pathname.search('/index') > 0) {
+
+            let mission = requestQuery.pathname.substring(1, requestQuery.pathname.search('/index'));
+
+            if (isNaN(mission)) {
+                return;
+            }
+
+            influxConnection.then(function (influx) {
+
+                influx.getMeasurements().then(names => {
+
+                    let query = `select * from`;
+                    names.forEach(name => {
+                        query += ` ${name},`;
+                    });
+                    query = query.slice(0, -1);
+                    query += ` where mission = '${mission}'`;
+
+                    influx.query(query).then(result => {
+
+                        let ids = {};
+                        result.forEach(measure => {
+                            if (!ids.hasOwnProperty(measure.id)) {
+                                ids[measure.id] = measure.time;
+                            }
+                        });
+
+                        response.setHeader("Access-Control-Allow-Origin", '*');
+                        response.end(JSON.stringify(ids));
+
+                    }).catch(function (err) {
+                        console.error(`Error querying data from InfluxDB! ${err.stack}`);
+
+                        if (WSS.readyState !== WebSocket.OPEN) {
+                            return;
+                        }
+
+                        WSS.send(id + ':error:' + e);
+                    });
+                }).catch(function (err) {
+                    console.error(`Error getting measurements! ${err.stack}`);
+
+                    if (WSS.readyState !== WebSocket.OPEN) {
+                        return;
+                    }
+
+                    WSS.send(id + ':error:' + e);
+                });
+            }).catch(function (err) {
+                console.error(`Error connecting to InfluxDB! ${err.stack}`);
+
+                if (WSS.readyState !== WebSocket.OPEN) {
+                    return;
+                }
+
+                WSS.send(id + ':error:' + e);
+            });
+
+        } else {
+
+            let mission = requestQuery.pathname.substring(1);
+
+            if (isNaN(mission)) {
+                return;
+            }
+
+            let ids = requestQuery.searchParams.getAll('ids[]');
+
+            influxConnection.then(function (influx) {
+
+                influx.getMeasurements().then(names => {
+
+                    let queries = [];
+                    ids.forEach(id => {
+
+                        let query = `select * from`;
+                        names.forEach(name => {
+                            query += ` ${name},`;
+                        });
+                        query = query.slice(0, -1);
+                        query += ` where (mission = '${mission}') and (id = '${id}')`;
```

Comment:
Whenever we make a query to a database using user input, we're vulnerable to something called SQL injection: a malicious user could pass in an id that deleted all our data. I won't go into the details of how it works -- you can read https://www.w3schools.com/sql/sql_injection.asp for that -- but we should be verifying all user input and returning an error (status code should be in the 400 range) if the mission or any of the ids are invalid.

Code to see if a mission is valid: `/^\d+$/.test(mission)`.
Code to see if a id is valid: `/^[0-9A-F]{8}-[0-9A-F]{4}-4[0-9A-F]{3}-[89AB][0-9A-F]{3}-[0-9A-F]{12}$/i.test(id)`.

These both use something called regular expressions. For the mission tester, it says make sure the mission string matches the format beginning of string (`^`), at least one (`+`) digit (`\d`) followed by the end of the string `$`. I got the regular expression for the second just by googling it: https://stackoverflow.com/questions/19989481/how-to-determine-if-a-string-is-a-valid-v4-uuid